compiler/tflchef/core/src/CustomOp/MaxPoolWithArgMax.cpp

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "MaxPoolWithArgMax.h"
+
+#include "flatbuffers/flexbuffers.h"
+
+flatbuffers::Offset<void> MaxPoolWithArgMaxChef::value(flatbuffers::FlatBufferBuilder &fbb) const
+{
+  return flatbuffers::Offset<void>();
+}
+
+flatbuffers::Offset<flatbuffers::Vector<uint8_t>>
+MaxPoolWithArgMaxChef::custom_value(flatbuffers::FlatBufferBuilder &fbb) const
+{
+  auto &operation = (*_operation);
+
+  assert(operation.type() == "MaxPoolWithArgMax");
+
+  auto flex_buffers = std::make_unique<flexbuffers::Builder>();
+  size_t map_start = flex_buffers->StartMap();
+
+  flex_buffers->Int("filter_width", operation.max_pool_with_argmax_options().filter_width());
+  flex_buffers->Int("filter_height", operation.max_pool_with_argmax_options().filter_height());
+  flex_buffers->Int("stride_w", operation.max_pool_with_argmax_options().stride_w());
+  flex_buffers->Int("stride_h", operation.max_pool_with_argmax_options().stride_h());
+  flex_buffers->Int("padding", operation.max_pool_with_argmax_options().padding());
+  flex_buffers->Int("activation", operation.max_pool_with_argmax_options().activation());
+  flex_buffers->Int("output_type", operation.max_pool_with_argmax_options().output_type());
```

Comment:
Do you have any criteria for adding attribute? The model containing `MaxPoolArgWithArgMax` that acquired from `res/TensorFlowPythonExamples` has custom values as follows.
![image](https://user-images.githubusercontent.com/24587354/96824697-6647c300-146a-11eb-917a-f3c4b737b806.png)

You can interpret its value with flex buffer method([like this](https://github.com/Samsung/ONE/blob/a0b5e3fa85d7140f8c36ba99687e313298a05ade/compiler/luci/pass/src/ResolveCustomOpBatchMatMulPass.cpp#L39)) or you can do it just simply by printing its ASCII code.
```
s,t,r,i,d,e,s,,,,,,,i,n,c,l,u,d,e,_,b,a,t,c,h,_,i,n,_,i,n,d,e,x,,k,s,i,z,e,,,,,,,p,a,d,d,i,n,g,,,V,A,L,I,D,,,2,,,B,,,,,,,@,h,,,,,,$,,
```
In any way, the information you saved is a little bit different from what it should be. Please let me know if there is any information that I am misunderstood.

Reply:
Thanks! I didn't find it in tensorflow for some reason.

Reply:
@d-krylov Actually, this is what I think it would be.
```suggestion
  auto start = flex_buffers->StartVector("ksize");
  flex_buffers->Add(1);
  flex_buffers->Add(operation.max_pool_with_argmax_options().filter_width());
  flex_buffers->Add(operation.max_pool_with_argmax_options().filter_height());
  flex_buffers->Add(1);
  flex_buffers->EndVector(start, /*typed=*/true, /*fixed=*/false);
  start = flex_buffers->StartVector("strides");
  flex_buffers->Add(1);
  flex_buffers->Add(operation.max_pool_with_argmax_options().stride_w());
  flex_buffers->Add(operation.max_pool_with_argmax_options().stride_h());
  flex_buffers->Add(1);
  flex_buffers->EndVector(start, /*typed=*/true, /*fixed=*/false);
  std::string padding = operation.max_pool_with_argmax_options().padding() ? "VALID":"SAME";
  flex_buffers->String("padding", padding);
  flex_buffers->Bool("include_batch_in_index",
                     operation.max_pool_with_argmax_options().include_batch_in_index());
```

In custom value, `output_type` and `activation` doesn't exist - I might be missing though. And When I cook the recipe you posted in #4721, the model has two outputs as of now. Maybe this should be fixed as well.
![image](https://user-images.githubusercontent.com/24587354/96856879-5d221a80-1499-11eb-822d-de5e898dccec.png)

Reply:
Thank you for that detailed explanation!

> In custom value, output_type and activation doesn't exist - I might be missing though

- "output_type" attribute is actually a "Targmax" value If I understand it correctly, need to rename
- "activation" attribute is a mistake, it was added because we decided to include it into circle/luci representation (similar to other operators, like conv, and other pools)

>  And When I cook the recipe you posted in #4721, the model has two outputs as of now

I am not sure why netron did not render two outputs (both are listed in details in the right panel in your [comment](https://github.com/Samsung/ONE/pull/4719/files#r509873656))

Reply:
> "output_type" attribute is actually a "Targmax" value If I understand it correctly, need to rename

Oh, I didn't know that. `output_type` looks good to me:)

> I am not sure why netron did not render two outputs (both are listed in details in the right panel in your comment)

Graph output is only one I think.
```
operand {
  name: "ifm"
  type: FLOAT32
  shape { dim: 1 dim: 18 dim: 18 dim: 1 }
}
operand {
  name: "ofm"
  type: FLOAT32
  shape { dim: 1 dim: 9 dim: 9 dim: 1 }
}
operand {
  name: "argmax"
  type: INT64
  shape { dim: 1 dim: 9 dim: 9 dim: 1 }
}
operation {
  type: "MaxPoolWithArgMax"
  input: "ifm"
  output: "ofm"
  output: "argmax"
  max_pool_with_argmax_options {
    padding: SAME
    filter_width: 1
    filter_height: 1
    stride_w: 2
    stride_h: 2
    activation: NONE
    output_type: INT64
    include_batch_in_index: false
  }
}
input: "ifm"
output: "ofm"
output: "argmax" # >> this should be removed
```

Well, I'm not sure whether the number of output of this model is one or two. I think two output looks more natural for me. As of now we can just go as it is and fix it when it becomes a problem:)

Reply:
Ok, I was confused by tflite custom operators format, but I think now I understand what is going on better.

To make things clear:
There are at least **TWO** different ways to save custom op in TFLite model:
1. Use [select operators](https://www.tensorflow.org/lite/guide/ops_select) technology.
Looks like this method serializes operator attributes Protocol Buffers format (wrapped in flexbuffers. Yes, proto inside flatbuffers).
I think this is what @mhs4670go used in his first post, so custom options are shown as a simple byte array. Netron can not parse this format.
2. Use raw custom operators.
This method serialize attributes in a flexbuffers Map.
We are using this format in our runtime and in this PR to store custom options.

I found that these two ways saves different sets of attributes and we can not decide what to store in (2) looking at (1).

Soooo, I looked carefully at tflite models generated with raw custom options and found that there are stored only 4 of them:
- include_batch_in_index
- ksize
- padding
- strides

<details>
code to generate select ops

```python
import tensorflow as tf
import numpy as np
root = tf.train.Checkpoint()
input_data = np.ones([2,3,4,5], dtype = np.float32)
root.f = tf.function(lambda x: tf.nn.max_pool_with_argmax(x,ksize=[1,1,1,1],strides=[1,1,1,1],padding='SAME',output_dtype=tf.dtypes.int64,include_batch_in_index=False))
concrete_func = root.f.get_concrete_function(input_data)
converter = tf.lite.TFLiteConverter.from_concrete_functions([concrete_func])

converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS,
                                       tf.lite.OpsSet.SELECT_TF_OPS]

tflite_model = converter.convert()
with open("MaxPoolWithArgMax.tflite", "wb") as f:
    f.write(tflite_model)

```

code to generate raw custom operators

```python
import tensorflow as tf
import numpy as np
root = tf.train.Checkpoint()
input_data = np.ones([2,3,4,5], dtype = np.float32)
root.f = tf.function(lambda x: tf.nn.max_pool_with_argmax(x,ksize=[1,1,1,1],strides=[1,1,1,1],padding='SAME',output_dtype=tf.dtypes.int64,include_batch_in_index=False))
concrete_func = root.f.get_concrete_function(input_data)
converter = tf.lite.TFLiteConverter.from_concrete_functions([concrete_func])

converter.allow_custom_ops=True

tflite_model = converter.convert()
with open("MaxPoolWithArgMax.tflite", "wb") as f:
    f.write(tflite_model)

```

</details>

Reply:
Thank you for details! Here is what I tested including what you mentioned.
1. use `res/res/TensorFlowPythonExamples/tfpem.py` with 2.X converter
```bash
# custom value
84, 0, 84, 97, 114, 103, 109, 97, 120, 0, 105, 110, 99, 108, 117, 100, 101, 95, 98, 97, 116, 99, 104, 95, 105, 110, 95, 105, 110, 100, 101, 120, 0, 107, 115, 105, 122, 101, 0, 4, 1, 2, 2, 1, 112, 97, 100, 100, 105, 110, 103, 0, 5, 86, 65, 76, 73, 68, 0, 115, 116, 114, 105, 100, 101, 115, 0, 4, 1, 1, 1, 1, 6, 73, 72, 65, 43, 33, 19, 6, 1, 6, 0, 4, 0, 45, 33, 19, 4, 4, 104, 44, 20, 44, 12, 36, 1
# tfldump
T(0) Targmax(4) include_batch_in_index(false) ksize((?)) padding(VALID) strides((?))
```
2. use `res/res/TensorFlowPythonExamples/tfpem.py` with 1.X converter (what I posted at first)
```bash
# custom value
115, 116, 114, 105, 100, 101, 115, 0, 4, 1, 1, 1, 1, 105, 110, 99, 108, 117, 100, 101, 95, 98, 97, 116, 99, 104, 95, 105, 110, 95, 105, 110, 100, 101, 120, 0, 107, 115, 105, 122, 101, 0, 4, 1, 2, 2, 1, 112, 97, 100, 100, 105, 110, 103, 0, 5, 86, 65, 76, 73, 68, 0, 4, 50, 28, 18, 66, 4, 1, 4, 0, 28, 16, 64, 104, 44, 20, 44, 8, 36, 1
# tfldump
include_batch_in_index(false) ksize((?)) padding(VALID) strides((?))
```
3. use select ops with 2.X
```bash
# custom value
17, 77, 97, 120, 80, 111, 111, 108, 87, 105, 116, 104, 65, 114, 103, 109, 97, 120, 0, 155, 18, 17, 77, 97, 120, 80, 111, 111, 108, 87, 105, 116, 104, 65, 114, 103, 109, 97, 120, 26, 0, 42, 7, 10, 1, 84, 18, 2, 48, 1, 42, 19, 10, 7, 115, 116, 114, 105, 100, 101, 115, 18, 8, 10, 6, 26, 4, 1, 1, 1, 1, 42, 28, 10, 22, 105, 110, 99, 108, 117, 100, 101, 95, 98, 97, 116, 99, 104, 95, 105, 110, 95, 105, 110, 100, 101, 120, 18, 2, 40, 0, 42, 13, 10, 7, 84, 97, 114, 103, 109, 97, 120, 18, 2, 48, 9, 42, 17, 10, 5, 107, 115, 105, 122, 101, 18, 8, 10, 6, 26, 4, 1, 1, 1, 1, 42, 17, 10, 7, 112, 97, 100, 100, 105, 110, 103, 18, 6, 18, 4, 83, 65, 77, 69, 50, 19, 10, 17, 77, 97, 120, 80, 111, 111, 108, 87, 105, 116, 104, 65, 114, 103, 109, 97, 120, 0, 2, 176, 158, 20, 20, 4, 40, 1
# tfldump
O(0:0) CUSTOM(FlexMaxPoolWithArgmax) 
    # NOTHING..
    I T(0:0) x
    O T(0:1) Identity
    O T(0:2) Identity_1
```
4. use raw custom op as you posted above with 2.X
```
# custom value
84, 0, 84, 97, 114, 103, 109, 97, 120, 0, 105, 110, 99, 108, 117, 100, 101, 95, 98, 97, 116, 99, 104, 95, 105, 110, 95, 105, 110, 100, 101, 120, 0, 107, 115, 105, 122, 101, 0, 4, 1, 1, 1, 1, 112, 97, 100, 100, 105, 110, 103, 0, 4, 83, 65, 77, 69, 0, 115, 116, 114, 105, 100, 101, 115, 0, 4, 1, 1, 1, 1, 6, 72, 71, 64, 42, 32, 19, 6, 1, 6, 0, 4, 0, 44, 32, 19, 4, 4, 104, 44, 20, 44, 12, 36, 1
# tfldump
T(0) Targmax(4) include_batch_in_index(false) ksize((?)) padding(SAME) strides((?))
```

> There is nothing about output type. This information embedded into output tensors in tflite.

I'm not sure where the information is located. As test results, number 4 case seems natural. It includes `Targmax` and has two graph outputs. we might as well follow this graph. How about your opinion?
![image](https://user-images.githubusercontent.com/24587354/97246947-f8bcdd80-1841-11eb-9e31-96150d8a7c0c.png)

Reply:
```cxx
 /**
   * REGISTER_OP("MaxPoolWithArgmax")
    .Attr("ksize: list(int) >= 4")
    .Attr("strides: list(int) >= 4")
    .Attr("Targmax: {int32, int64} = DT_INT64")
    .Attr(GetPaddingAttrString())
    .Attr("include_batch_in_index: bool = false")
    .Input("input: T")
    .Output("output: T")
    .Output("argmax: Targmax")
    .Attr("T: realnumbertype")
    .SetShapeFn([](InferenceContext* c) {
      TF_RETURN_IF_ERROR(shape_inference::MaxPoolShape(c));
      c->set_output(1, c->output(0));
      return Status::OK();
    });
   */
```
I mean, the `Attr` part should be custom value - ksize, strides, Targmax, Padding, include_batch_in_index, T